### PR TITLE
Implemented feature/favorites-icon 

### PR DESCRIPTION
**Pull request by Dmytro1991ua**
Create a favorite icon (btn) component which will take a certain movie id and add it to local storage and remove it on click to an icon. An icon will be changed depending on whether local storage is empty or not. Also added additional methods to our API handlers ( in order to properly work with local storage ) which allow checking whether there is data in favorites and history on a local storage

### CODE REVIEW

File: src/components/FavoritesBtn.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { useState } from 'react';
+
+import { FavoriteMovie } from '../models/index';
+import api from '../service/api';
+
+import '../styles/FavoritesBtn.scss';
+
+const FavoritesBtn: React.FC<FavoriteMovie> = ({ movieId }) => {
+  const [icon, setIcon] = useState(api.isIdInFavorites(movieId));
```
Comment (oleksandrstarov) [inline src/components/FavoritesBtn.tsx:9]: What is a type of icon?

File: src/components/FavoritesBtn.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  const heartUnfilledIcon = '/images/favoriteBtn/heartFilled.svg';
+  const heartFilledIcon = '/images/favoriteBtn/heartUnfilled.svg';
+
+  const handleClick = (event: React.MouseEvent<HTMLElement>): void => {
+    event.preventDefault();
+    api.isIdInFavorites(movieId)
```
Comment (oleksandrstarov) [inline src/components/FavoritesBtn.tsx:16]: why not use icon?

File: src/components/FavoritesBtn.tsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      ? api.setFavoritesId(movieId)
+      : api.deleteFavoritsId(movieId);
+    setIcon(api.isIdInFavorites(movieId));
+  };
+
+  return (
+    <a href="#!" className="favorite-btn" onClick={handleClick}>
+      <img
+        className="favorite-icon"
+        src={icon ? heartFilledIcon : heartUnfilledIcon}
```
Comment (oleksandrstarov) [inline src/components/FavoritesBtn.tsx:26]: is icon boolean type? if yes, choose more appropriate name

File: src/models/index.tsx [hunk 4]
```diff
@@ -15,3 +15,8 @@ export enum Theme {
   dark,
   light,
 }
+
+export interface FavoriteMovie {
+  movieId: number;
+}
```
Comment (oleksandrstarov) [inline src/models/index.tsx:21]: I dont see any reason why we need this to be interface. Seems like FavoritesBtnProps type will be enough

**Review by oleksandrstarov — CHANGES_REQUESTED** [review]
Check comments

File: src/components/FavoritesBtn.tsx (hunk 1 above)
Comment (Dmytro1991ua) [reply src/components/FavoritesBtn.tsx:9]: The type of  icon is  a boolean

File: src/components/FavoritesBtn.tsx (hunk 2 above)
Comment (Dmytro1991ua) [reply src/components/FavoritesBtn.tsx:16]: You're right. I have not noticed it at the beginning. I am going to use a state variable here

File: src/components/FavoritesBtn.tsx (hunk 3 above)
Comment (Dmytro1991ua) [reply src/components/FavoritesBtn.tsx:26]: I agree that the name isn't appropriate. Here, in a state, I would like to store a boolean value that allows to check local storage and based on that data to change the state of a favorite icon. I am thinking of names like favoriteIconState, setFavoriteIconState. is it an appropriate name?

File: src/models/index.tsx (hunk 4 above)
Comment (Dmytro1991ua) [reply src/models/index.tsx:21]: Thanks for a hint. Did you mean just to use a type FavoritesBtnProps directly in a component?

File: src/service/api.tsx [hunk 5]
```diff
@@ -19,35 +19,43 @@ const api = {
...
     return store.theme;
   },
 
   setFavoritesId(id: number) {
     const store = this.getStore();
     const isThereAnId = store.favorites.includes(id);
-    if (isThereAnId) {
-      store.favorites.push(id)
-    };
+    if (!isThereAnId) {
```
Comment (solomiia-sahala) [inline src/service/api.tsx:47]: please, describe changes of this line in description as well

**Review by solomiia-sahala — APPROVED** [review]

File: src/service/api.tsx (hunk 5 above)
Comment (Dmytro1991ua) [reply src/service/api.tsx:47]: Here an isThereAnId variable holds information either there is any id in favorites on local storage or not(boolean value). Current changes allow to correctly check data in local storage and eventually will use to  add or delete id from local storage as well as change a state of favorite icon

File: src/service/api.tsx [hunk 6]
```diff
@@ -120,7 +137,7 @@ const api = {
 
   changeImgLinks(url: string, size: string) {
     return `${API_IMG_URL}${size}${url}`;
-  },
-}
+  }
+};
 
 export default api;
```
Comment (Kachyr) [inline src/service/api.tsx:143]: add newline

File: src/components/FavoritesBtn.tsx (hunk 3 above)
Comment (Kachyr) [reply src/components/FavoritesBtn.tsx:26]: maybe you can use name like - "iconStatus" or something similar

File: src/components/FavoritesBtn.tsx (hunk 1 above)
Comment (Kachyr) [reply src/components/FavoritesBtn.tsx:9]: you can set type for "setState" as useState<boolean>(api.isIdInFavorites(movieId));

**Review by Kachyr — APPROVED** [review]
LGTM, check comments

File: src/components/FavoritesBtn.tsx (hunk 1 above)
Comment (Dmytro1991ua) [reply src/components/FavoritesBtn.tsx:9]: Thanks, i'll have a look at this one

File: src/service/api.tsx (hunk 6 above)
Comment (Dmytro1991ua) [reply src/service/api.tsx:143]: done

File: src/components/FavoritesBtn.tsx (hunk 3 above)
Comment (Dmytro1991ua) [reply src/components/FavoritesBtn.tsx:26]: Good one, but I still believe that I need to stay consistent in terms of  working with favorites icon, hence I have a feeling that favoriteIconState could be more suitable here

Comment (oleksandrstarov) [reply src/components/FavoritesBtn.tsx:26]: as a rule boolean variable should start with is prefix, here the most appropriate name from my point of view is isFavorite

File: src/models/index.tsx (hunk 4 above)
Comment (oleksandrstarov) [reply src/models/index.tsx:21]: Yes

File: src/service/api.tsx (hunk 5 above)
Comment (oleksandrstarov) [reply src/service/api.tsx:47]: maybe rename to isNewId

File: src/components/FavoritesBtn.tsx (hunk 3 above)
Comment (Dmytro1991ua) [reply src/components/FavoritesBtn.tsx:26]: Thanks for a hint. I have already renamed it.

File: src/models/index.tsx (hunk 4 above)
Comment (Dmytro1991ua) [reply src/models/index.tsx:21]: Thanks

File: src/service/api.tsx (hunk 5 above)
Comment (Dmytro1991ua) [reply src/service/api.tsx:47]: already changed

File: src/models/index.tsx [hunk 7]
```diff
@@ -15,3 +14,5 @@ export enum Theme {
   dark,
   light,
 }
+
+
```
Comment (oleksandrstarov) [inline src/models/index.tsx:18]: ah

File: src/service/api.tsx [hunk 8]
```diff
@@ -19,35 +19,43 @@ const api = {
...
+    if (!isNewId) {
+      store.favorites.push(id);
+    }
     this.setStore(store);
   },
 
+  isIdInFavorites(id: number) {
+    const store = api.getFavoritsIdList();
+    const isNewId = store.includes(id);
+    return !isNewId;
```
Comment (oleksandrstarov) [inline src/service/api.tsx:56]: why not just return !store.favorites.includes(id)

File: src/service/api.tsx [hunk 9]
```diff
@@ -67,13 +75,19 @@ const api = {
...
+    if (!isNewId) {
+      store.history.push(id);
+    }
     this.setStore(store);
   },
 
+  isIdInHistory(id: number) {
+    const store = api.getHistoryIdList();
+    const isNewId = store.includes(id);
+    return !isNewId;
```
Comment (oleksandrstarov) [inline src/service/api.tsx:88]: same as above

File: src/styles/FavoritesBtn.scss [hunk 10]
```diff
@@ -0,0 +1,18 @@
+@import "../styles/variables.scss";
+
```
Comment (oleksandrstarov) [inline src/styles/FavoritesBtn.scss:2]: Check indent please

**Review by oleksandrstarov — APPROVED** [review]
Check comments,
Looks good

File: src/service/api.tsx [hunk 11]
```diff
@@ -93,9 +107,10 @@ const api = {
   getDataByIds(ids: number[]) {
     const urls = ids.map((id: number) => `movie/${id}?${API_KEY}`);
     const requests = urls.map(
-      async (url: any) => await axios.get(url).then((res: { data: any; }) => res.data)
+      async (url: any) =>
```
Comment (AnnaBolshak) [inline src/service/api.tsx:110]: Why "any"? does url have string type?

Comment (AnnaBolshak) [reply src/service/api.tsx:110]: Is it possible to describe type for "data" instead of any?

File: src/service/api.tsx [hunk 12]
```diff
@@ -67,13 +75,19 @@ const api = {
 
   setHistoryId(id: number) {
     const store = this.getStore();
-    const isThereAnId = store.history.includes(id);
-    if (isThereAnId) {
-      store.history.push(id)
-    };
+    const isNewId = store.history.includes(id);
+    if (!isNewId) {
+      store.history.push(id);
```
Comment (AnnaBolshak) [inline src/service/api.tsx:80]: It seems like here can be used approach from comment of @oleksandrstarov too (!store.favorites.includes(id))

File: src/components/FavoritesBtn.tsx [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+  const heartFilledIcon = '/images/favoriteBtn/heartUnfilled.svg';
+
+  const handleClick = (event: React.MouseEvent<HTMLElement>): void => {
+    event.preventDefault();
+    isFavorite ? api.setFavoritesId(movieId) : api.deleteFavoritsId(movieId);
+    setisFavorite(api.isIdInFavorites(movieId));
+  };
+
+  return (
+    <a href="#!" className="favorite-btn" onClick={handleClick}>
```
Comment (AnnaBolshak) [inline src/components/FavoritesBtn.tsx:24]: Could you explain, why using <a> is needed here?

File: src/components/FavoritesBtn.tsx [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+import api from '../service/api';
+
+import '../styles/FavoritesBtn.scss';
+
+type FavoritesBtnProps = {
+  movieId: number;
+};
+
+const FavoritesBtn: React.FC<FavoritesBtnProps> = ({ movieId }) => {
+  const [isFavorite, setisFavorite] = useState(api.isIdInFavorites(movieId));
```
Comment (AnnaBolshak) [inline src/components/FavoritesBtn.tsx:12]: Please check the name of setisFavorite - starting from the second, each word should start with a capital letter

File: src/service/api.tsx [hunk 15]
```diff
@@ -120,7 +137,7 @@ const api = {
 
   changeImgLinks(url: string, size: string) {
```
Comment (AnnaBolshak) [inline src/service/api.tsx:138]: it seems like this function does not change the link but return link (usually prefix "get" is used)

File: src/components/FavoritesBtn.tsx [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+  const heartUnfilledIcon = '/images/favoriteBtn/heartFilled.svg';
+  const heartFilledIcon = '/images/favoriteBtn/heartUnfilled.svg';
+
+  const handleClick = (event: React.MouseEvent<HTMLElement>): void => {
```
Comment (AnnaBolshak) [inline src/components/FavoritesBtn.tsx:17]: Please use more specific name for the function

File: src/components/FavoritesBtn.tsx (hunk 13 above)
Comment (AnnaBolshak) [inline src/components/FavoritesBtn.tsx:24]: Is using of the `<a>` here needed?

Comment (Dmytro1991ua) [reply src/components/FavoritesBtn.tsx:24]: Here, I wanted to be more specific semantically, that's why decided to wrap an Icon in <a> tag

File: src/components/FavoritesBtn.tsx (hunk 14 above)
Comment (Dmytro1991ua) [reply src/components/FavoritesBtn.tsx:12]: Thanks for a hint. I have not noticed it at the first place

File: src/models/index.tsx (hunk 7 above)
Comment (Dmytro1991ua) [reply src/models/index.tsx:18]: fixed

File: src/styles/FavoritesBtn.scss (hunk 10 above)
Comment (Dmytro1991ua) [reply src/styles/FavoritesBtn.scss:2]: fixed

File: src/components/FavoritesBtn.tsx (hunk 16 above)
Comment (Dmytro1991ua) [reply src/components/FavoritesBtn.tsx:17]: In the beginning, I believe I wanted to create a more generic name, but now it actually makes more sense that the name should be more specific. Is handleFavoritesIconClick appropriate?

File: src/service/api.tsx (hunk 11 above)
Comment (Dmytro1991ua) [reply src/service/api.tsx:110]: thanks for pointing it out. Changed types for 'url' and 'data'

File: src/service/api.tsx (hunk 15 above)
Comment (Dmytro1991ua) [reply src/service/api.tsx:138]: thanks for pointing it out. I'll have a look at it

File: src/service/api.tsx (hunk 8 above)
Comment (Dmytro1991ua) [reply src/service/api.tsx:56]: fixed

File: src/service/api.tsx (hunk 9 above)
Comment (Dmytro1991ua) [reply src/service/api.tsx:88]: fixed

File: src/service/api.tsx (hunk 12 above)
Comment (Dmytro1991ua) [reply src/service/api.tsx:80]: fixed

**Pull request merged**